File: core/src/test/java/javabot/admin/JavadocTest.java
package javabot.admin;

import com.jayway.awaitility.Awaitility;
import com.jayway.awaitility.Duration;
import javabot.dao.ApiDao;
import javabot.dao.EventDao;
import javabot.dao.JavadocClassDao;
import javabot.javadoc.JavadocApi;
import javabot.model.ApiEvent;
import javabot.model.EventType;
import javabot.operations.BaseOperationTest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testng.Assert;
import org.testng.annotations.Test;

import javax.inject.Inject;
import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.util.concurrent.TimeUnit;

@Test
public class JavadocTest extends BaseOperationTest {
    private static final Logger LOG = LoggerFactory.getLogger(JavadocTest.class);

    @Inject
    private ApiDao apiDao;

    @Inject
    private JavadocClassDao javadocClassDao;

    @Inject
    private EventDao eventDao;

    @Test
    public void servlets() throws IOException {
        String apiName = "Servlet";
        dropApi(apiName);
        addApi(apiName, "http://tomcat.apache.org/tomcat-7.0-doc/servletapi/",
               "https://repo1.maven.org/maven2/javax/servlet/javax.servlet-api/3.0.1/javax.servlet-api-3.0.1.jar");
        checkServlets(apiName);
    }

    @Test(dependsOnMethods = "servlets")
    public void reloadServlets() {
        String apiName = "Servlet";
        ApiEvent event = new ApiEvent(EventType.RELOAD, getTestUser().getNick(), apiDao.find(apiName).getId());
        eventDao.save(event);
        waitForEvent(event, "reloading " + apiName, new Duration(30, TimeUnit.MINUTES));
        getMessages().get();
        checkServlets(apiName);
    }

    private void checkServlets(final String apiName) {
        Assert.assertEquals(
                               javadocClassDao.getClass(apiDao.find(apiName), "javax.servlet.http", "HttpServletRequest").length, 1);
        scanForResponse("~javadoc HttpServlet", "javax.servlet.http.HttpServlet");
        scanForResponse("~javadoc HttpServlet.doGet(*)", "javax.servlet.http.HttpServlet.doGet");
        scanForResponse("~javadoc HttpServletRequest", "javax.servlet.http.HttpServletRequest");
        scanForResponse("~javadoc HttpServletRequest.getMethod()", "javax.servlet.http.HttpServletRequest.getMethod");
    }

    @Test
    public void javaee() throws IOException {
        String apiName = "JavaEE7";
        dropApi(apiName);
        addApi(apiName, "http://docs.oracle.com/javaee/7/api/", "https://repo1.maven.org/maven2/javax/javaee-api/7.0/javaee-api-7.0.jar");
        scanForResponse("~javadoc Annotated", "javax.enterprise.inject.spi.Annotated");
        scanForResponse("~javadoc Annotated.getAnnotation(*)", "javax.enterprise.inject.spi.Annotated.getAnnotation");
        scanForResponse("~javadoc ContextService", "javax.enterprise.concurrent.ContextService");
        scanForResponse("~javadoc ContextService.createContextualProxy(*)", "createContextualProxy(Object, Class...)");
        scanForResponse("~javadoc ContextService.createContextualProxy(*)",
                        "createContextualProxy(Object, Map<String, String>, Class...)");
        scanForResponse("~javadoc ContextService.createContextualProxy(*)", "createContextualProxy(T, Class<T>)");
        scanForResponse("~javadoc ContextService.createContextualProxy(*)",
                        "createContextualProxy(T, Map<String, String>, Class<T>)");
        scanForResponse("~javadoc PartitionPlan", "[JavaEE7: javax.batch.api.partition.PartitionPlan]");
        scanForResponse("~javadoc PartitionPlan.setPartitionProperties(Properties[])",
                        "javax.batch.api.partition.PartitionPlan.setPartitionProperties(Properties[])");
    }

    @Test
    public void jdk() throws MalformedURLException {
        getJavabot();
        if (Boolean.valueOf(System.getProperty("dropJDK", "false"))) {
            System.out.println("Dropping JDK API");
            dropApi("JDK");
            System.out.println("Done");
        }
        JavadocApi api = apiDao.find("JDK");
        if (api == null) {
            ApiEvent event = new ApiEvent(getTestUser().getNick(), "JDK", "http://docs.oracle.com/javase/8/docs/api",
                                          new File(System.getProperty("java.home"), "lib/rt.jar").toURI().toURL().toString());
            eventDao.save(event);
            waitForEvent(event, "adding JDK", new Duration(30, TimeUnit.MINUTES));
            getMessages().get();
            api = apiDao.find("JDK");
        }
        Assert.assertEquals(javadocClassDao.getClass(api, "java.lang", "Integer").length, 1);
    }

    private void addApi(final String apiName, final String apiUrlString, final String downloadUrlString) {
        ApiEvent event = new ApiEvent(getTestUser().getNick(), apiName, apiUrlString, downloadUrlString);
        eventDao.save(event);
        waitForEvent(event, "adding " + apiName, new Duration(30, TimeUnit.MINUTES));
        LOG.info("done waiting for event to finish");
        getMessages().get();
    }

    private void dropApi(final String apiName) {
        getJavabot();
        final ApiEvent event = new ApiEvent(EventType.DELETE, getTestUser().getNick(), apiName);
        eventDao.save(event);
        waitForEvent(event, "dropping " + apiName, Duration.FIVE_MINUTES);
        Awaitility.await()
                  .atMost(60, TimeUnit.SECONDS)
                  .until(() -> apiDao.find(apiName) == null);
        getMessages().get();
    }
}

File: src/test/kotlin/javabot/admin/JavadocTest.kt
package javabot.admin

import jakarta.inject.Inject
import java.io.File
import javabot.BaseTest
import javabot.JavabotConfig
import javabot.dao.JavadocClassDao
import javabot.model.ApiEvent
import javabot.model.javadoc.JavadocApi
import javabot.operations.JavadocOperation
import org.testng.Assert
import org.testng.Assert.assertNotNull
import org.testng.Assert.fail
import org.testng.annotations.BeforeClass
import org.testng.annotations.Test

@Test
class JavadocTest() : BaseTest() {
    companion object {
        val eeVersion = "8.0.0"
        val eeApiName = "JakartaEE8"
    }

    @Inject private lateinit var classDao: JavadocClassDao

    @Inject private lateinit var config: JavabotConfig

    @Inject private lateinit var operation: JavadocOperation

    @BeforeClass
    fun drops() {
        apiDao.delete(eeApiName)
    }

    private fun checkServlets(api: JavadocApi) {
        assertNotNull(
            classDao.getClass(api, "jakarta.servlet.http", "HttpServletRequest"),
            "Should find an entry for ${api.name}/jakarta.servlet.http.HttpServletRequest"
        )

        scanForResponse(
            operation.handleMessage(message("~javadoc HttpServlet")),
            "javax/servlet/http/HttpServlet.html"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc HttpServlet.doGet(*)")),
            "javax/servlet/http/HttpServlet.html#doGet"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc HttpServletRequest")),
            "javax/servlet/http/HttpServletRequest.html"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc HttpServletRequest.getMethod()")),
            "javax/servlet/http/HttpServletRequest.html#getMethod"
        )
        checkServletFile()
    }

    private fun checkServletFile() {
        val uri = File("javadoc/$eeApiName/$eeVersion/javax/servlet/http/HttpServlet.html")
        Assert.assertTrue(uri.exists())
    }

    @Test(dependsOnMethods = ["core"])
    fun jakartaEE() {
        val api = loadApi(eeApiName, "jakarta.platform", "jakarta.jakartaee-api", "8.0.0")
        verifyMapCount()
        scanForResponse(
            operation.handleMessage(message("~javadoc Annotated")),
            "javax/enterprise/inject/spi/Annotated.html"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc Annotated.getAnnotation(*)")),
            "javax/enterprise/inject/spi/Annotated.html#getAnnotation"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc ContextService")),
            "javax/enterprise/concurrent/ContextService.html"
        )

        scanForResponse(
            operation.handleMessage(message("~javadoc ContextService.createContextualProxy(*)")),
            "createContextualProxy-java.lang.Object-java.lang.Class...-"
        )

        scanForResponse(
            operation.handleMessage(message("~javadoc ContextService.createContextualProxy(*)")),
            "createContextualProxy-java.lang.Object-java.util.Map-java.lang.Class...-"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc ContextService.createContextualProxy(*)")),
            "createContextualProxy-T-java.lang.Class-"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc ContextService.createContextualProxy(*)")),
            "createContextualProxy-T-java.util.Map-java.lang.Class-"
        )

        checkServlets(api)
    }

    @Test
    fun core() {
        val api = loadApi("JDK", version = "11")
        Assert.assertEquals(classDao.getClass(api, "Map").size, 1)
        assertNotNull(
            classDao.getClass(api, "java.lang", "Integer"),
            "Should find an entry for ${api.name}'s java.lang.Integer"
        )
        assertNotNull(
            classDao.getClass(api, "java.util", "List"),
            "Should find an entry for ${api.name}'s java.util.List"
        )
        assertNotNull(
            classDao.getClass(api, "java.sql", "ResultSet"),
            "Should find an entry for ${api.name}'s java.sql.ResultSet"
        )
        assertNotNull(
            classDao.getClass(api, "java.util.Map.Entry"),
            "Should find an entry for ${api.name}'s java.util.Map.Entry"
        )
        assertNotNull(
            classDao.getClass(api, "Map.Entry"),
            "Should find an entry for ${api.name}'s java.util.Map.Entry"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc Map.Entry")),
            "${api.baseUrl}/java.base/java/util/Map.Entry.html"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc String.chars()")),
            "${api.baseUrl}/java.base/java/lang/String.html#chars()"
        )
        scanForResponse(
            operation.handleMessage(message("~javadoc ResultSet.getInt(*)")),
            "${api.baseUrl}/java.sql/java/sql/ResultSet.html#getInt"
        )
    }

    private fun verifyMapCount() {
        val list = classDao.getClass(null, "Map")
        Assert.assertEquals(list.size, 1, "Should have found only 1:  $list")
    }

    @Test(dependsOnMethods = ["core", "jakartaEE"])
    fun guava() {
        val apiName = "guava"
        val guava = loadApi(apiName, "com.google.guava", "guava", "28.2-jre")
        val classCount = classDao.count()

        Assert.assertEquals(classDao.getClass(guava, "ArrayTable").size, 1)
        Assert.assertEquals(classDao.getClass(guava, "AbstractCache").size, 1)
        Assert.assertEquals(classDao.getClass(guava, "ArrayBasedCharEscaper").size, 1)
        Assert.assertEquals(classDao.getClass(null, "ArrayList").size, 1)
        Assert.assertEquals(classDao.getClass(null, "HttpServlet").size, 1)

        classDao.delete(classDao.getClass(guava, "AbstractCache")[0])
        classDao.delete(classDao.getClass(guava, "ArrayTable")[0])

        Assert.assertEquals(classDao.count(), classCount - 2)

        Assert.assertEquals(classDao.getClass(null, "ArrayList").size, 1)
        Assert.assertEquals(classDao.getClass(null, "HttpServlet").size, 1)

        Assert.assertEquals(classDao.getClass(guava, "AbstractCache").size, 0)
        Assert.assertEquals(classDao.getClass(guava, "ArrayTable").size, 0)

        val event = ApiEvent.reload(TEST_USER.nick, apiName)
        //        injector.injectMembers(event)
        fail()
        event.handle()

        Assert.assertEquals(classDao.getClass(event.api, "AbstractCache").size, 1)
        Assert.assertEquals(classDao.getClass(event.api, "ArrayTable").size, 1)
        Assert.assertEquals(classDao.getClass(null, "ArrayList").size, 1)
        Assert.assertEquals(classDao.getClass(null, "HttpServlet").size, 1)
    }
}
